_posts/2023-06-22-a-simple-rest-api-with-play-framework-and-scala.md

```diff
@@ -0,0 +1,420 @@
+
+```scala
+package repositories
+
+import persistence.Persistence
+import persistence.entities.Car
+import repositories.errors.RepositoryError
+
+import scala.collection.mutable.ListBuffer
+
+class CarRepository(val db: Persistence[Car, Long, ListBuffer]){
+  def findById(id: Long): Either[RepositoryError, Option[Car]] = handleIfErrors(db.find(id))
+  def save(car: Car): Either[RepositoryError, Car] = handleIfErrors(db.insert(car))
+  def update(car: Car): Either[RepositoryError, Unit] = handleIfErrors(db.update(car))
+  def delete(id: Long): Either[RepositoryError, Unit] = handleIfErrors(db.delete(id))
+  def findAll(): Either[RepositoryError, List[Car]] = handleIfErrors(db.findAll())
+
+  private def handleIfErrors[A](f: => A) = {
+    try{
+      Right(f)
+    } catch {
+      case e: RuntimeException => Left(RepositoryError(e.getMessage))
+    }
+  }
+}
+```
+## 3.2 Managers
+On this layer we define our managers. Usually containing logic to manage other resources like repositories, transactions, connections, security etc and contains business logic (here is too simple) of our application.
+Here we define a simple manager for car entities:
+![alt "Managers structure"](../images/play-rest-api/managers_structure.jpeg)
+```scala
+package managers
+
+import managers.errors.ManagerError
+import models.CarDTO
+import repositories.CarRepository
+import models.util.CarConversions._
+import repositories.errors.RepositoryError
+
+trait Manager{
+  def get(id: Long): Either[ManagerError, Option[CarDTO]]
+  def save(car: CarDTO): Either[ManagerError, CarDTO]
+  def update(car: CarDTO): Either[ManagerError, Unit]
+  def delete(car: CarDTO): Either[ManagerError, Unit]
+  def getAll(): Either[ManagerError, List[CarDTO]]
+}
+
+class CarManager(repository: CarRepository) extends Manager {
+  override def get(id: Long): Either[ManagerError, Option[CarDTO]] =
+    repository
+      .findById(id)
+      .fold(error => Left(toManagerError(error)), car => Right(car.map(_.toDTO())))
+
+
+
+  override def save(carDTO: CarDTO): Either[ManagerError, CarDTO] =
+    repository
+      .save(carDTO.toDB())
+      .fold(error => Left(toManagerError(error)), car => Right(car.toDTO()))
+
+
+  override def update(carDTO: CarDTO): Either[ManagerError, Unit] =
+    repository
+      .update(carDTO.toDB())
+      .fold(error => Left(toManagerError(error)), _ => Right(()))
+
+  override def delete(carDTO: CarDTO): Either[ManagerError, Unit] =
+    repository
+      .delete(carDTO.id.getOrElse(0L))
+      .fold(error => Left(toManagerError(error)), _ => Right(()))
+
+  override def getAll(): Either[ManagerError, List[CarDTO]] =
+    repository
+      .findAll()
+      .fold(error => Left(toManagerError(error)), carList => Right(carList.map(_.toDTO())))
+
+  private def toManagerError(repositoryError: RepositoryError) = ManagerError(repositoryError.message)
+
+}
+```
+
+## 3.3 Controllers
+This is the layer where we define ouρ APIs public endpoints.
```

Comment:
Need to explain the structure of an action - how it works, what it's made of. You can take one action and break it down and explain every step. All actions are then explainable